source/extensions/access_loggers/wasm/config.cc

```diff
@@ -29,22 +29,39 @@ WasmAccessLogFactory::createAccessLogInstance(const Protobuf::Message& proto_con
   auto vm_id = config.config().vm_config().vm_id();
   auto root_id = config.config().root_id();
   auto configuration = std::make_shared<std::string>(config.config().configuration());
-  auto tls_slot = context.threadLocal().allocateSlot();
-  // Create a base WASM to verify that the code loads before setting/cloning the for the
-  // individual threads.
-  auto plugin = std::make_shared<Common::Wasm::Plugin>(
-      config.config().name(), config.config().root_id(), config.config().vm_config().vm_id(),
-      envoy::api::v2::core::TrafficDirection::UNSPECIFIED, context.localInfo(),
-      nullptr /* listener_metadata */, context.scope());
-  auto base_wasm =
-      Common::Wasm::createWasm(config.config().vm_config(), plugin, context.clusterManager(),
-                               context.dispatcher(), context.api());
-  // NB: the Slot set() call doesn't complete inline, so all arguments must outlive this call.
-  tls_slot->set([base_wasm, configuration](Event::Dispatcher& dispatcher) {
-    return std::static_pointer_cast<ThreadLocal::ThreadLocalObject>(
-        Common::Wasm::getOrCreateThreadLocalWasm(*base_wasm, *configuration, dispatcher));
-  });
-  return std::make_shared<WasmAccessLog>(root_id, std::move(tls_slot), std::move(filter));
+  auto access_log = std::make_shared<WasmAccessLog>(root_id, nullptr, std::move(filter));
+
+  auto callback = [access_log, &context, &config, configuration](const std::string& code) {
+    auto tls_slot = context.threadLocal().allocateSlot();
+    // Create a base WASM to verify that the code loads before setting/cloning the for the
+    // individual threads.
+    auto plugin = std::make_shared<Common::Wasm::Plugin>(
```

Comment:
If `code` is not included in the `Plugin`, we can. I added `code` in the `Plugin` in this PR so we cannot. But we can keep `code` out of the `Plugin`. Which one would you prefer?

Reply:
Can you pass it into Common::Wasm::createWasm().  The plugin is supposed to have only information specific to the particular plugin point, while the code is only used to create the initial Wasm (or non-cloneable thread-local Wasm(s)) but in any case is common to all plugin points.   The new PR I submitted uses the copy constructor of Wasm for all the thread-local versions of the VMs, so you can store the code in Wasm and then access it if necessary in the copy constructor, but *do not store it* in the copy constructed Wasm(s) because it could be big and we don't want to waste the memory.  The code in the base_wasm will be deleted when the last thread-local Wasm is created.

Reply:
@jplevyak update the PR to pass the `code` to `Common::Wasm::createWasm()` instead of putting it in `Plugin`